spotbugs/etc/messages.xml

```diff
@@ -9128,6 +9137,51 @@ Using floating-point variables should not be used as loop counters, as they are
       ]]>
         </Details>
   </BugPattern>
+
+  <BugPattern type="ENV_USE_PROPERTY_INSTEAD_OF_ENV">
+    <ShortDescription>It is preferable to use portable Java property instead of environment variable.</ShortDescription>
+    <LongDescription>It is preferable to use portable Java property '{3}' instead of environment variable '{2}' in method {1}.</LongDescription>
+    <Details>
+      <![CDATA[<p>
+        Environment variables are not portable, the variable name itself (not only the value) may be different depending on the running OS.
+        Not only the names of the specific environment variables can differ (e.g. `USERNAME` in Windows and `USER` in Unix systems),
+        but even the semantics differ, e.g. the case sensitivity (Windows being case-insensitive and Unix case-sensitive).
+        Thus, the environment variables may not obey the general contract of the <code>Object.equals(java.lang.Object)</code> and <code>Object.hashCode()</code> methods.
```

Comment:
>  Thus, the environment variables may not obey the general contract of the <code>Object.equals(java.lang.Object)</code> and <code>Object.hashCode()</code> methods.

This sound wrong. The contract of mentioned methods is for sure working for String values read from environment variables.

Reply:
However, it sounds counterintuitive, I think it is right according to the following sources:

This part from `java.lang.System.getenv(String)`: 
> On UNIX systems the alphabetic case of name is typically significant, while on Microsoft Windows systems it is typically not. For example, the expression System.getenv("FOO").equals(System.getenv("foo")) is likely to be true on Microsoft Windows.

This part from `java.lang.System.getenv()`:
> The returned map and its collection views may not obey the general contract of the [Object.equals(java.lang.Object)](https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#equals-java.lang.Object-) and [Object.hashCode()](https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#hashCode--) methods.

Do you think I misunderstood these parts from the Java API? Or was my wording misunderstandable?

Reply:
I seems to me that the potential issue about equals/hashcode is with the map itself, not with the keys and values.
So you might run into issues doing something like `someMap.put("env", System.getEnv())`.

The current message implies that the values (or keys) inside the map do not implement equals/hashcode according to the general contract. My understanding is that `someMap.put("env", System.getEnv("whatever"))` is fine

Reply:
Thank you both for the feedback. I modified the message. Could you please look at it one more time?